apps/src/music/progress/ProgressManager.ts

```diff
@@ -0,0 +1,129 @@
+// This file contains a generic ProgressManager which any lab can include,
+// if it wants to make progress without reloading the page.
+
+// Abstract class that validates a set of conditions. How
+// the validation works is up to the implementor.
+export abstract class Validator {
+  abstract shouldCheckConditions(): boolean;
+  abstract checkConditions(): void;
+  abstract conditionsMet(conditions: string[]): boolean;
+  abstract clear(): void;
+}
+
+// A validation inside the progression.
+interface Validation {
+  conditions: string[];
+  message: string;
+  next: boolean;
+}
+
+// The definition of the progression.
+interface Progression extends Object {
```

Comment:
does this need to extend `Object`?

Reply:
Good catch; this was left over from before I added full details.